File: ui_playwright_bdd/scripts/agents/runner-agent.js
'use strict';
// ============================================================
// Runner Agent — Exécution des tests Playwright/Cucumber
// ============================================================
// Absorbe : execute-agent, qa-agent (run), smoke, regression, flaky
//
// Usage:
//   node scripts/agents/runner-agent.js run              Lance tous les tests
//   node scripts/agents/runner-agent.js smoke            Tests @smoke uniquement
//   node scripts/agents/runner-agent.js critical         Tests @critical uniquement
//   node scripts/agents/runner-agent.js regression       Détecte les régressions vs baseline
//   node scripts/agents/runner-agent.js flaky [--runs=N] Lance N fois, identifie les tests instables
//   node scripts/agents/runner-agent.js baseline         Sauvegarde le baseline actuel
//
// Coût LLM : très bas — LLM utilisé uniquement pour le résumé narratif final
// ============================================================
require('dotenv').config();
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

const fs     = require('fs');
const path   = require('path');
const { spawnSync } = require('child_process');
const llm    = require('./llm');
const tracer = require('./shared/tracer');
const memory = require('./shared/memory-store');

const FRAMEWORK    = path.join(__dirname, '..', '..');
const RESULTS_DIR  = path.join(FRAMEWORK, 'allure-results');
const DOCS_DIR     = path.join(FRAMEWORK, 'docs');
const BASELINE_FILE = path.join(DOCS_DIR, 'baseline.json');

fs.mkdirSync(DOCS_DIR, { recursive: true });

const DRY_RUN  = process.argv.includes('--dry-run');
const RUNS_ARG = parseInt(process.argv.find(a => a.startsWith('--runs='))?.split('=')[1] || '3');
const FLAKY_THRESHOLD = 0.34; // 1/3 des runs = flaky

const G = '\x1b[32m', R = '\x1b[31m', Y = '\x1b[33m', C = '\x1b[36m', B = '\x1b[1m', E = '\x1b[0m';

// ── Lecture allure-results ─────────────────────────────────────────────────────
function loadResults() {
  if (!fs.existsSync(RESULTS_DIR)) return [];
  return fs.readdirSync(RESULTS_DIR)
    .filter(f => f.endsWith('-result.json'))
    .map(f => { try { return JSON.parse(fs.readFileSync(path.join(RESULTS_DIR, f), 'utf8')); } catch { return null; } })
    .filter(Boolean);
}

function computeStats(results) {
  const stats = { passed: 0, failed: 0, broken: 0, skipped: 0, total: 0 };
  for (const r of results) {
    const s = r.status || 'unknown';
    if (s in stats) stats[s]++;
    stats.total++;
  }
  stats.pass_rate = stats.total ? Math.round(stats.passed / stats.total * 1000) / 10 : 0;
  return stats;
}

function printStats(stats, label = '') {
  const bar  = '\x1b[32m' + '█'.repeat(Math.round(stats.pass_rate / 5)) + '\x1b[0m' + '░'.repeat(20 - Math.round(stats.pass_rate / 5));
  console.log(`\n  [${bar}] ${stats.pass_rate}% passed  ${label}`);
  console.log(`  Total: ${stats.total}  |  ${G}${stats.passed}P${E}  ${R}${stats.failed}F${E}  ${Y}${stats.broken}B${E}  ${C}${stats.skipped}S${E}`);
}

// ── Exécution cucumber-js ──────────────────────────────────────────────────────
function runCucumber(tags = '') {
  const args = [
    'cucumber-js',
    'src/features/**/*.feature',
    '--require-module', 'ts-node/register',
    '--require', 'src/core/world.ts',
    '--require', 'src/hooks/**/*.ts',
    '--require', 'src/steps/**/*.ts',
    '--format', 'allure-cucumberjs/reporter',
    '--format', 'progress',
  ];
  if (tags) args.push('--tags', tags);

  console.log(`  ${C}→ npx ${args.join(' ')}${E}`);
  if (DRY_RUN) { console.log(`  ${Y}[DRY-RUN] Tests non exécutés${E}`); return 0; }

  const result = spawnSync('npx', args, {
    cwd: FRAMEWORK, stdio: 'inherit', shell: true, encoding: 'utf8'
  });
  return result.status || 0;
}

// ── Résumé LLM ────────────────────────────────────────────────────────────────
async function llmSummary(stats, failures) {
  const span = new tracer.Span('runnerSummary', JSON.stringify(stats), llm.MODEL).begin();
  try {
    const prompt = `Tu es un expert QA. Génère un résumé court (2-3 phrases) du run de tests Playwright.

Résultats : ${stats.passed}/${stats.total} passés (${stats.pass_rate}%)
Échecs : ${failures.map(f => f.name).join(', ') || 'aucun'}

Sois factuel et direct. Si PASSED → rassurant. Si FAILED → cause probable + action.`;

    const resp = await llm.chat([{ role: 'user', content: prompt }]);
    const summary = resp.message.content || '';
    span.response = summary; span.end(true);
    return summary;
  } catch (e) {
    span.error = e.message; span.end(false);
    return `Run terminé : ${stats.pass_rate}% de réussite.`;
  }
}

// ── run : tous les tests ───────────────────────────────────────────────────────
async function cmdRun() {
  console.log(`\n${B}=== RUNNER — RUN [${llm.MODEL}] ===${E}`);
  const code = runCucumber();
  const results  = loadResults();
  const stats    = computeStats(results);
  const failures = results.filter(r => ['failed','broken'].includes(r.status));
  printStats(stats);

  const summary = await llmSummary(stats, failures);
  console.log(`\n  ${C}${summary}${E}`);

  memory.recordEpisode('runner-agent', failures.map(f => ({ tc: f.name, status: f.status })), summary, 'run', stats);
  if (code !== 0 && !DRY_RUN) process.exitCode = 1;
}

// ── smoke : tests @smoke ──────────────────────────────────────────────────────
async function cmdSmoke() {
  console.log(`\n${B}=== RUNNER — SMOKE ===${E}`);
  const code = runCucumber('@smoke');
  const results = loadResults();
  const stats   = computeStats(results.filter(r => (r.labels||[]).some(l => l.value === 'smoke')));
  printStats(stats, '@smoke');
  if (code !== 0 && !DRY_RUN) process.exitCode = 1;
}

// ── critical : tests @critical ────────────────────────────────────────────────
async function cmdCritical() {
  console.log(`\n${B}=== RUNNER — CRITICAL ===${E}`);
  const code = runCucumber('@critical');
  const results = loadResults();
  printStats(computeStats(results), '@critical');
  if (code !== 0 && !DRY_RUN) process.exitCode = 1;
}

// ── flaky : N runs pour détecter les tests instables ─────────────────────────
async function cmdFlaky() {
  console.log(`\n${B}=== RUNNER — FLAKY DETECTION (${RUNS_ARG} runs) ===${E}`);
  const tcMap = {}; // tc_name → { fails, total }

  for (let i = 1; i <= RUNS_ARG; i++) {
    console.log(`\n  ${C}Run ${i}/${RUNS_ARG}${E}`);
    runCucumber();
    const results = loadResults();
    for (const r of results) {
      const name = r.name || r.fullName || 'unknown';
      if (!tcMap[name]) tcMap[name] = { fails: 0, total: 0 };
      tcMap[name].total++;
      if (['failed','broken'].includes(r.status)) tcMap[name].fails++;
    }
  }

  const flaky = Object.entries(tcMap)
    .map(([name, s]) => ({ name, flakiness: s.fails / s.total, fails: s.fails, runs: s.total }))
    .filter(t => t.flakiness > 0 && t.flakiness < 1)
    .sort((a, b) => b.flakiness - a.flakiness);

  const critical = flaky.filter(t => t.flakiness >= FLAKY_THRESHOLD);

  console.log(`\n  ${B}Tests flaky détectés : ${flaky.length}${E}`);
  for (const t of flaky) {
    const icon = t.flakiness >= FLAKY_THRESHOLD ? R + '⚠ FLAKY' + E : Y + '~ instable' + E;
    console.log(`  ${icon}  ${t.name.slice(0,55).padEnd(55)} ${Math.round(t.flakiness*100)}% (${t.fails}/${t.runs})`);
  }

  if (!DRY_RUN) {
    fs.writeFileSync(path.join(DOCS_DIR, 'flaky-report.json'), JSON.stringify({ ts: new Date().toISOString(), runs: RUNS_ARG, flaky, critical }, null, 2), 'utf8');
    console.log(`\n  ${G}✓ docs/flaky-report.json${E}`);
  }

  memory.recordEpisode('runner-agent', critical.map(t => ({ tc: t.name, flakiness: t.flakiness })), `${critical.length} tests flaky critiques`, 'flaky');
}

// ── regression : compare avec baseline ────────────────────────────────────────
async function cmdRegression() {
  console.log(`\n${B}=== RUNNER — REGRESSION ===${E}`);
  const baseline = fs.existsSync(BASELINE_FILE) ? JSON.parse(fs.readFileSync(BASELINE_FILE, 'utf8')) : null;
  if (!baseline) {
    console.log(`  ${Y}⚠  Aucun baseline — lance d'abord: runner baseline${E}`);
  }

  runCucumber();
  const results = loadResults();
  const stats   = computeStats(results);
  printStats(stats, 'vs baseline');

  if (baseline) {
    const regressions = results.filter(r => {
      const prev = baseline.tests?.[r.name];
      return ['failed','broken'].includes(r.status) && prev === 'passed';
    });
    if (regressions.length) {
      console.log(`\n  ${R}⚠  ${regressions.length} régression(s) détectée(s) :${E}`);
      regressions.forEach(r => console.log(`    ${R}✗${E} ${r.name}`));
    } else {
      console.log(`\n  ${G}✓ Aucune régression${E}`);
    }
  }
}

// ── baseline : sauvegarde l'état actuel ───────────────────────────────────────
function cmdBaseline() {
  console.log(`\n${B}=== RUNNER — BASELINE ===${E}`);
  const results = loadResults();
  const tests   = {};
  results.forEach(r => { tests[r.name] = r.status; });
  const baseline = { ts: new Date().toISOString(), total: results.length, tests };
  if (!DRY_RUN) {
    fs.writeFileSync(BASELINE_FILE, JSON.stringify(baseline, null, 2), 'utf8');
    console.log(`  ${G}✓ Baseline sauvegardé : ${results.length} tests${E}`);
  } else {
    console.log(`  ${Y}[DRY-RUN] Baseline non sauvegardé (${results.length} tests)${E}`);
  }
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function main() {
  const [cmd] = process.argv.slice(2).filter(a => !a.startsWith('--'));
  await llm.assertRunning();

  switch (cmd) {
    case 'run':        await cmdRun();        break;
    case 'smoke':      await cmdSmoke();      break;
    case 'critical':   await cmdCritical();   break;
    case 'flaky':      await cmdFlaky();      break;
    case 'regression': await cmdRegression(); break;
    case 'baseline':   cmdBaseline();         break;
    default:
      console.log(`
${B}Runner Agent${E} — Exécution des tests Playwright/Cucumber

  run              Lance tous les tests + résumé LLM
  smoke            Lance uniquement les tests @smoke
  critical         Lance uniquement les tests @critical
  regression       Détecte les régressions vs baseline
  flaky            Détecte les tests instables (multi-runs)
  baseline         Sauvegarde le baseline actuel

Options:
  --dry-run        Simulation sans exécution ni écriture
  --runs=N         Nombre de runs pour la détection flaky (défaut: 3)
`);
  }
}

main().catch(e => { console.error(R + e.message + E); process.exit(1); });

File: api-Java-Rest-Assured/agents/runner-agent.py
# ============================================================
# Runner Agent — Exécution des tests API RestAssured (Maven)
# ============================================================
# Commandes :
#   python agents/runner-agent.py run [env]         → tous les tests
#   python agents/runner-agent.py smoke [env]       → @smoke uniquement
#   python agents/runner-agent.py critical [env]    → @critical
#   python agents/runner-agent.py regression [env]  → @regression
#   python agents/runner-agent.py flaky [--runs=3]  → détection flaky
#   python agents/runner-agent.py baseline          → sauvegarde baseline
#   python agents/runner-agent.py gono-go           → run + quality gate
# ============================================================

import sys, os, json, glob, subprocess, time, argparse
sys.stdout.reconfigure(encoding="utf-8", errors="replace")
sys.path.insert(0, os.path.dirname(__file__))

import llm

FRAMEWORK  = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
TARGET_DIR = os.path.join(FRAMEWORK, "target")
ALLURE_DIR = os.path.join(TARGET_DIR, "allure-results")
DOCS_DIR   = os.path.join(FRAMEWORK, "docs")
BASELINE   = os.path.join(DOCS_DIR, "baseline.json")

R = "\033[31m"; G = "\033[32m"; Y = "\033[33m"; C = "\033[36m"
W = "\033[1m";  E = "\033[0m"

QUALITY_GATE = {"pass_rate": 90.0, "fail_rate": 5.0}


# ── Maven ────────────────────────────────────────────────────────────────────

MVN = "mvn.cmd" if sys.platform == "win32" else "mvn"


def mvn(args: list, env_name: str = "local") -> int:
    cmd = [MVN, "clean", "test", f"-Denv={env_name}"] + args
    print(f"\n{C}  mvn {' '.join(args)}{E}")
    result = subprocess.run(cmd, cwd=FRAMEWORK)
    return result.returncode


def mvn_tag(tag: str, env_name: str = "local") -> int:
    return mvn([f'-Dcucumber.filter.tags={tag}'], env_name)


# ── Lecture résultats Allure ───────────────────────────────────────────────────

def load_results() -> list:
    results = []
    for f in glob.glob(os.path.join(ALLURE_DIR, "*.json")):
        try:
            with open(f, encoding="utf-8") as fh:
                data = json.load(fh)
            if "name" in data and "status" in data:
                results.append(data)
        except Exception:
            pass
    return results


def parse_stats(results: list) -> dict:
    total   = len(results)
    passed  = sum(1 for r in results if r.get("status") == "passed")
    failed  = sum(1 for r in results if r.get("status") == "failed")
    broken  = sum(1 for r in results if r.get("status") == "broken")
    skipped = sum(1 for r in results if r.get("status") == "skipped")
    pass_rate = round(passed / total * 100, 1) if total else 0
    fail_rate = round((failed + broken) / total * 100, 1) if total else 0
    return {
        "total": total, "passed": passed, "failed": failed,
        "broken": broken, "skipped": skipped,
        "pass_rate": pass_rate, "fail_rate": fail_rate,
    }


def print_stats(stats: dict):
    gate_ok = stats["pass_rate"] >= QUALITY_GATE["pass_rate"] and stats["fail_rate"] <= QUALITY_GATE["fail_rate"]
    color   = G if gate_ok else R
    print(f"\n  {W}Résultats :{E}")
    print(f"  Total   : {stats['total']}")
    print(f"  {G}Passés  : {stats['passed']}{E}")
    print(f"  {R}Échoués : {stats['failed']}{E}")
    print(f"  {Y}Brisés  : {stats['broken']}{E}")
    print(f"  {color}Pass rate : {stats['pass_rate']}%{E}  |  Fail rate : {stats['fail_rate']}%")
    print(f"  {color}Quality Gate : {'PASS ✓' if gate_ok else 'FAIL ✗'}{E}")


# ── Commandes ──────────────────────────────────────────────────────────────────

def cmd_run(env: str = "local"):
    print(f"\n{W}RUNNER — Tous les tests (env={env}){E}")
    rc = mvn([], env)
    results = load_results()
    if results:
        print_stats(parse_stats(results))
    return rc


def cmd_smoke(env: str = "local"):
    print(f"\n{W}RUNNER — Smoke tests @smoke (env={env}){E}")
    rc = mvn_tag("@smoke", env)
    results = load_results()
    if results:
        print_stats(parse_stats(results))
    return rc


def cmd_critical(env: str = "local"):
    print(f"\n{W}RUNNER — Tests critiques @critical (env={env}){E}")
    rc = mvn_tag("@critical", env)
    results = load_results()
    if results:
        print_stats(parse_stats(results))
    return rc


def cmd_regression(env: str = "local"):
    print(f"\n{W}RUNNER — Régression (env={env}){E}")
    rc = mvn([], env)
    results = load_results()
    if results:
        print_stats(parse_stats(results))
    return rc


def cmd_flaky(runs: int = 3, env: str = "local"):
    print(f"\n{W}RUNNER — Détection flaky ({runs} runs, env={env}){E}")
    history = {}
    for i in range(runs):
        print(f"\n  {C}Run {i+1}/{runs}{E}")
        mvn([], env)
        for r in load_results():
            name = r.get("name", "?")
            history.setdefault(name, []).append(r.get("status", "unknown"))

    flaky = {n: statuses for n, statuses in history.items() if len(set(statuses)) > 1}
    os.makedirs(DOCS_DIR, exist_ok=True)
    with open(os.path.join(DOCS_DIR, "flaky-report.json"), "w", encoding="utf-8") as f:
        json.dump({"runs": runs, "flaky": flaky, "timestamp": time.strftime("%Y-%m-%dT%H:%M:%S")}, f, indent=2, ensure_ascii=False)

    if flaky:
        print(f"\n  {R}{len(flaky)} test(s) flaky :{E}")
        for name, statuses in flaky.items():
            print(f"  {Y}[{'/'.join(statuses)}]{E} {name[:65]}")
        flaky_text = "\n".join(f"- {n}: {' / '.join(s)}" for n, s in list(flaky.items())[:10])
        messages = [{"role": "user", "content": (
            f"Ces tests API RestAssured sont flaky ({runs} runs) :\n{flaky_text}\n\n"
            f"En 3 points, explique les causes probables (API publique partagée, timing, données) "
            f"et propose des actions de stabilisation."
        )}]
        print(f"\n{W}  Analyse LLM :{E}")
        try:
            for line in llm.chat(messages).strip().split("\n"):
                print(f"  {line}")
        except Exception as e:
            print(f"  {Y}LLM indisponible : {e}{E}")
    else:
        print(f"  {G}Aucun test flaky détecté sur {runs} runs.{E}")
    return flaky


def cmd_baseline():
    print(f"\n{W}RUNNER — Sauvegarde baseline{E}")
    results = load_results()
    if not results:
        print(f"  {R}Aucun résultat Allure — lance d'abord un run.{E}")
        return
    stats = parse_stats(results)
    os.makedirs(DOCS_DIR, exist_ok=True)
    baseline = {
        "timestamp": time.strftime("%Y-%m-%dT%H:%M:%S"),
        "stats": stats,
        "tests": [{"name": r["name"], "status": r["status"]} for r in results],
    }
    with open(BASELINE, "w", encoding="utf-8") as f:
        json.dump(baseline, f, indent=2, ensure_ascii=False)
    print(f"  {G}Baseline sauvegardée ({stats['total']} tests, {stats['pass_rate']}% pass).{E}")
    print(f"  Fichier : docs/baseline.json")


def cmd_gonogo(env: str = "local"):
    print(f"\n{W}RUNNER — Go/No-Go (run + gate, env={env}){E}")
    cmd_run(env)
    results = load_results()
    if not results:
        print(f"  {R}Aucun résultat.{E}")
        return False
    stats = parse_stats(results)
    gate_ok = stats["pass_rate"] >= QUALITY_GATE["pass_rate"] and stats["fail_rate"] <= QUALITY_GATE["fail_rate"]
    if gate_ok:
        print(f"\n  {G}{W}GO — Déploiement autorisé{E}")
    else:
        print(f"\n  {R}{W}NO-GO — Déploiement bloqué{E}")
        if stats["pass_rate"] < QUALITY_GATE["pass_rate"]:
            print(f"  {R}Pass rate {stats['pass_rate']}% < 90%{E}")
        if stats["fail_rate"] > QUALITY_GATE["fail_rate"]:
            print(f"  {R}Fail rate {stats['fail_rate']}% > 5%{E}")
    return gate_ok


# ── CLI ────────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Runner Agent — API RestAssured Maven")
    parser.add_argument("command", choices=["run", "smoke", "critical", "regression",
                                             "flaky", "baseline", "gono-go"])
    parser.add_argument("env", nargs="?", default="local",
                        choices=["local", "staging", "production"])
    parser.add_argument("--runs", type=int, default=3)
    args = parser.parse_args()

    cmd = args.command
    env = args.env

    if cmd == "run":          cmd_run(env)
    elif cmd == "smoke":      cmd_smoke(env)
    elif cmd == "critical":   cmd_critical(env)
    elif cmd == "regression": cmd_regression(env)
    elif cmd == "flaky":      cmd_flaky(args.runs, env)
    elif cmd == "baseline":   cmd_baseline()
    elif cmd == "gono-go":    cmd_gonogo(env)
